securityscanutils/commands/format_results.go

```diff
@@ -0,0 +1,257 @@
+	var (
+		versions             []string
+		stableOnlyConstraint *semver.Constraints
+		err                  error
+	)
+	minVersionToScan := opts.minScannedVersion
+	if minVersionToScan == "" {
+		log.Println("MinScannedVersion flag not set, scanning all versions from repo")
+	} else {
+		stableOnlyConstraint, err = semver.NewConstraint(fmt.Sprintf(">= %s", minVersionToScan))
+		if err != nil {
+			log.Fatalf("Invalid constraint version: %s", minVersionToScan)
+		}
+	}
+
+	for _, release := range releases {
+		// ignore beta releases when display security scan results
+		test, err := semver.NewVersion(release.GetTagName())
+		if err != nil {
+			continue
+		}
+		if stableOnlyConstraint == nil || stableOnlyConstraint.Check(test) {
+			versions = append(versions, test.String())
+		}
+	}
+	return versions
+}
+
+func BuildSecurityScanReportForRepo(tags []string, opts *formatResultsOptions) error {
+	// tags are sorted by minor version
+	latestTag := tags[0]
+	prevMinorVersion, _ := semver.NewVersion(latestTag)
+	for ix, tag := range tags {
+		semver, err := semver.NewVersion(tag)
+		if err != nil {
+			return err
+		}
+		if ix == 0 || semver.Minor() != prevMinorVersion.Minor() {
+			fmt.Printf("\n***Latest %d.%d.x %s Release: %s***\n\n", semver.Major(), semver.Minor(), opts.targetRepoWritten, tag)
+			err = printImageReportForRepo(tag, opts)
+			if err != nil {
+				return err
+			}
+			prevMinorVersion = semver
+		} else {
+			fmt.Printf("<details><summary> Release %s </summary>\n\n", tag)
+			err = printImageReportForRepo(tag, opts)
+			if err != nil {
+				return err
+			}
+			fmt.Println("</details>")
+		}
+	}
+
+	return nil
+}
+
+func printImageReportForRepo(tag string, opts *formatResultsOptions) error {
+	for _, image := range opts.allImages {
+		fmt.Printf("**%s %s image**\n\n", opts.targetRepoWritten, image)
+		url := "https://storage.googleapis.com/solo-gloo-security-scans/" + opts.targetRepo + "/" + tag + "/" + image + "_cve_report.docgen"
+		report, err := GetSecurityScanReport(url)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("%s\n\n", report)
+	}
+	return nil
+}
+
+func GetSecurityScanReport(url string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	var report string
+	if resp.StatusCode == http.StatusOK {
+		bodyBytes, _ := ioutil.ReadAll(resp.Body)
+		report = string(bodyBytes)
+	} else if resp.StatusCode == http.StatusNotFound {
+		// Older releases may be missing scan results
+		report = "No scan found\n"
+	}
+	resp.Body.Close()
+
+	return report, nil
+}
+
+// Reads in a file, and tries to turn it into a map from version constraints to lists of images
+// As a byproduct, it also caches all unique images found into the option field 'allImages'
+// Also, I'm not sure why I didn't just use a csv reader... oh well.
+func readImageVersionConstraintsFile(opts *formatResultsOptions) (map[string][]string, error) {
```

Comment:
I think you have some nearly duplicate functions in here? readImageVersionConstraintsFile and GetImagesPerVersionFromFile

Reply:
Would it be worth it to refactor?

Reply:
I just copied the code as is. My goal was to isolate commands into their own file, and I feared that refactoring the code while also moving it would be harder to verify. Given that the code works as is, I don't think its worth it to refactor yet